### ISSUE TITLE
Allow choosing of Steam download server

### ISSUE THREAD

**Issue by Xorinaes** [otavepto/gbe_fork#87]
I wanted to ask if there's perhaps a possibility to add an option of choosing steam download region? As the newest Granblue game seems to use it for matchmaking and that causes problems when trying to play multiplayer using the emu, so I thought that it might be because right now, there is no region assigned in the emu and that might be throwing the game for a loop?

**Comment by Detanup01** [comment]
what?
GB using US as default. but dont think it has anything to do with that, you can even override that

**Comment by Xorinaes** [comment]
Sorry, I'm not sure if I understand; on Steam version of the game, GBR uses the download region setting on Steam to matchmake. This is why so many people recommend changing it to Tokyo, as it's the most populated one. What does "it uses US as default" mean in this context? Like, is it a fallback or something? And how do you override it?

**Comment by Detanup01** [comment]
Can you write the AppId down for that game?

Refernecing to this:

https://github.com/otavepto/gbe_fork/blob/2cbc7864305afb0838d317708b9f06a3dfcb6698/dll/dll/steam_utils.h#L83

https://github.com/otavepto/gbe_fork/blob/2cbc7864305afb0838d317708b9f06a3dfcb6698/post_build/README.release.md?plain=1#L468

Also no idea what you trying with the "download region" since Goldberg only have this as region and no stuff like "download region"

**Comment by Xorinaes** [comment]
AppID is 881020, and that's why I'm asking if it'd be possible to implement it; By the "Download Region" I mean the first option when you navigate to Steam -> Settings -> Download.
![maxresdefault](https://github.com/otavepto/gbe_fork/assets/73740928/5e289e72-e01f-46ee-ad39-ed9b4be14b59)

**Comment by Xorinaes** [comment]
Yeah, that's the weirdest thing; you're right about it being only the download server and I understand your confusion, but it is what Granblue Fantasy Relink uses to matchmake. It was the same in God Eater 3 at one point. There's lots of posts on Steam recommending people to change this setting to Tokyo - Japan to get matches. In the first version of the game, if there were two friends who had this set to different regions, they could not play together, even if they invited themselves to a private match.

**Comment by Xorinaes** [comment]
Here's the promised log, I launched the game, started the lobby and got my friend to try to join; then, my friend started the lobby and I tried to join. Hopefully it helps you guys. I tried to scrub it from personal data (like changing my actual ip to "my_ip" or my friend's ip to "my_friend_ip_1", stuff like that, but I could've missed something.
[STEAM_LOG.zip](https://github.com/otavepto/gbe_fork/files/14551169/STEAM_LOG.zip)

**Comment by Xorinaes** [comment]
Unfortunately, I tried that release and every combo of those two config files (with neither one, with either one alone and with two of them at the same time) and it didn't work. Just to make sure, I also used the same file with another game that we are sure works properly (GTFO in this case) and it did work. I also tried release version, experimental version, steamclient with original steam_api and experimental version of steamclient, and no dice. Also tried with experimental disable_lan_only and without it. And using virtual LAN service (self-hosted SoftEther with me being the host) and well, nothing worked. I guess there's gotta be some obscure API or a weird way of doing things because it's only the second time that I recall that using download region as matchmaking server happened (with God Eater 3 being the first one), and I don't think Goldberg emu was around during that time.

**Comment by Xorinaes** [comment]
Hello, it's me again. The game got an update which was supposed to fix something regarding the matchmaking, so I gave it a shot again and no dice. I also used the newe
st version of emu from 16.03. I am attaching a new log file; it might help some, as there might've been some changes.

[STEAM_LOG.zip](https://github.com/otavepto/gbe_fork/files/14624229/STEAM_LOG.zip)

**Comment by Xorinaes** [comment]
I also found this script which supposedly fixes region lock in Monster Hunter Rise, which apparently worked/works the same way as Granblue Relink, with the matchmaking being dependent on the download region.

[region_lock_fix.txt](https://github.com/otavepto/gbe_fork/files/14625426/region_lock_fix.txt)

**Comment by Xorinaes** [comment]
No problem, I also thought of one thing - perhaps allowing us to set value of pszRegion would work? I am not sure if it's even possible or valid, but it's my closest guess. Otherwise, I really have no idea, I will continue my search for maybe something that could help. It would seem that this way of doing things, i.e. matchmaking by download region, is actually kinda common for Japanese ports of their games. Also, for some reason, Duck Game did it as well before some update in, I think, 2020-ish.

**Comment by Xorinaes** [comment]
I come with new findings - I have tested online-fixes emu for a bit, and it seems the game makes two connections that are not even attempted when using this emu. There is one for, what I presume, to be a news source - 
![obraz](https://github.com/otavepto/gbe_fork/assets/73740928/a902776f-9a6c-4766-a238-c823d83bca55) - 
this connection appeared every time I tried to access the "News" section in the game, and the news worked correctly. When using gbe, a simple "Network Error has occurred" is displayed, and no connection attempt seems to be made (at least I couldn't see one in logs and in TCP View) - 
![obraz](https://github.com/otavepto/gbe_fork/assets/73740928/281998bc-db48-4b21-8d7d-eca33c984ee0).

The other, more crucial connection is what I think is the actual matchmaking server (at least for my region, I haven't tried testing it with other download regions). This appeared every time I first connected to online mode during a given play session - 
![obraz](https://github.com/otavepto/gbe_fork/assets/73740928/f6893cd9-f8ac-46a9-b472-64f01ab9d31e).

The ports were different every time, but the IP was consistent in both cases.

Additionally, when connecting to online services using online-fix, the game makes a multitude of connections on a bunch of different ports - 
![obraz](https://github.com/otavepto/gbe_fork/assets/73740928/bb81270e-871c-43e6-b645-7dd6568c785f) - behaviour not observed when using gbe (refer to above screenshot where only 4 connections are made using gbe). I dunno it it'll help you guys in any way, but it might mean that it's simply not possible to get it working with gbe, as it would seem that the game also relies on some outside servers which, for some reason, gbe doesn't even attempt to connect to? I, of course, have the "disable_lan_only" option enabled, and "offline" and "disable_networking" options disabled. Or is there something I can do to force the emu to connect to those IPs?

**Comment by Xorinaes** [comment]
A test mod that might have something to do with the issue was just created: https://github.com/Nenkai/gbfr.fix.matchmaking

It seems to hook into ELobbyDistanceFilter. You guys might be able to get some more info from that.

**Comment by cardinalsad** [comment]
I also got the "Network Connection Error" issue when entering the lobby, and I also found that when the network is connected, if the host logs out, I can join successfully.

**Comment by Xorinaes** [comment]
> I also found that when the network is connected, if the host logs out, I can join successfully.

Not quite sure that I follow; were you ultimately able to play together?

**Comment by cardinalsad** [comment]
> > I also found that when the network is connected, if the host logs out, I can join successfully.
> 
> 
> 
> Not quite sure that I follow; were you ultimately able to play together?

No, because you have to exit one to enter the other. In my log, it seems that the room has been added, and then the host can also see in the game that there are currently two people.

**Comment by cardinalsad** [comment]
I also saw the multiple udp connections under online fix that I observed in previous comments. Those connections are suddenly created when the host creates a room.
